### PR TITLE
Fix memory leaks reported by ASAN (with suppressions turned off) part 1

### LINKED ISSUE #10476: memory leak on foreign key constraint failure
Repoducer:

```lua
local fiber = require('fiber')
box.cfg{}
local x = box.schema.create_space('x', {format = {
    {'id', 'unsigned'},
    {'ref', 'string'},
}})
x:create_index('pk')
x:create_index('sk', {parts = {'ref'}})
local a = string.rep('a', 512 * 1024)
x:insert({1, a})
local y = box.schema.create_space('y', {format = {
    {'id', 'unsigned'},
    {'ref', 'string', foreign_key = {space = 'x', field = 'ref'}},
}})
y:create_index('pk')
y:create_index('sk', {parts = {'ref'}})
y:insert({1, a})
local b = string.rep('b', 512 * 1024)
while true do
    pcall(x.replace, x, {1, b})
    fiber.yield()
end
```

The script is killed by OOM killer after several seconds after start:
```
авг 26 17:23:45 mynote kernel: Out of memory: Killed process 93403 (tarantool) total-vm:21728684kB, anon-rss:10900736kB, file-rss:3396kB, shmem-rss:0kB, UID:1000 pgtables:41380kB oom_score_adj:0
```

### LINKED ISSUE #10477: Memory leak in iproto override triggers
Found by ASAN.

Reproducer:

```lua
local net = require('net.box')
local trigger = require('trigger')

box.cfg{listen = 3301}
box.schema.user.grant('guest', 'super', nil, nil, {if_not_exists = true})
trigger.set('box.iproto.override.eval', '1', function()
    return false
end)

local conn = net.connect(3301)
local s = string.rep('x', 512 * 1024)
while true do
    conn:eval('return 1', {s})
end
```

The script is killed by OOM killer after few dozens of seconds after start:
```
авг 26 17:55:03 mynote kernel: Out of memory: Killed process 94131 (tarantool) total-vm:20579488kB, anon-rss:10399196kB, file-rss:600kB, shmem-rss:0kB, UID:1000 pgtables:39060kB oom_score_adj:0
```

Introduced in `3.1.0-entrypoint-193-g6980007160`.

### LINKED ISSUE #10479: Memory leak on xlog open failure
Found by ASAN.

Reproducer:

```lua
local xlog = require('xlog')
while true do
    -- echo 1 > bad.xlog
    pcall(xlog.pairs, 'bad.xlog')
end
```

It takes less then a minute for the script to be killed by OOM killer:
```
авг 27 10:47:59 mynote kernel: Out of memory: Killed process 104668 (tarantool) total-vm:10361296kB, anon-rss:10295976kB, file-rss:96kB, shmem-rss:0kB, UID:1000 pgtables:20324kB oom_score_adj:0
```

Introduced in `1.7.2-148-g7e44ebd91c`.

### LINKED ISSUE #10480: Memory leak on replica disconnection from origin
Found by ASAN.

Reproducer.

Start a master:
```lua
local fiber = require('fiber')
box.cfg{
    memtx_max_tuple_size = 128 * 1024 * 1024,
    memtx_memory = 1024 * 1024 * 1024,
    replication_timeout = 10,
    work_dir = 'box-a',
    listen = 3301,
    read_only = false,
}
box.schema.user.grant('guest', 'super', nil, nil, {if_not_exists = true})
local space = box.schema.create_space('test')
space:create_index('pk')
while true do
    space:replace({1, string.rep(s, 128 * 1024 * 1024 - 100)})
    fiber.sleep(0.1)
    collectgarbage()
end
```

Start a replica:
```lua
local fiber = require('fiber')
box.cfg{
    memtx_max_tuple_size = 128 * 1024 * 1024, -- we need large tuples to make applier leak fast
    memtx_memory = 1024 * 1024 * 1024,
    replication_timeout = 10, -- default does not allow transmitting very large tuples
    work_dir = 'box-b',
    replication = 'localhost:3301',
    read_only = true,
}
while true do
    box.cfg{replication = {}}
    fiber.sleep(1) -- let master's relay finish disconnecting from this replica
    local lsn = box.info().replication[1].lsn
    box.cfg{replication = 'localhost:3301'}
    -- Make sure we read a tuple from master before disconnecting, so the leak is of tuple size
    while box.info().replication[1].lsn == lsn do
        fiber.sleep(0.01)
    end
end
```

The replica leak noticeably (I wait until it reached about 3 Gb).

Introduced in `2.10.0-beta2-53-gdacbf708f4`.

### LINKED ISSUE #10482: Memory leak in in `coio_connect_timeout`
Found by ASAN.

It is easier to demonstrate using ASAN. Reproducer:

```lua
local net = require('net.box')
net.connect(777)
```

```
 $ ./build-asan-debug/src/tarantool test.lua

=================================================================
==437649==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 76 byte(s) in 1 object(s) allocated from:
    #0 0x5a240d91c349 in malloc (/home/shiny/dev/tarantool/build-asan-debug/src/tarantool+0x1256349) (BuildId: cf86baf56ac08c441fd7e98ccd34d38f9f70bc00)
    #1 0x7f65acf55767 in generate_addrinfo /usr/src/debug/glibc/glibc/nss/getaddrinfo.c:1081:16
    #2 0x7f65acf55767 in gaih_inet /usr/src/debug/glibc/glibc/nss/getaddrinfo.c:1199:12
    #3 0x7f65acf55767 in getaddrinfo /usr/src/debug/glibc/glibc/nss/getaddrinfo.c:2391:12
    #4 0x5a240d8d38d9 in getaddrinfo.part.0 asan_interceptors.cpp.o
    #5 0x5a240ed8fccb in getaddrinfo_cb /home/shiny/dev/tarantool/src/lib/core/coio_task.c:327:13
    #6 0x5a240ed8a06a in coio_on_feed /home/shiny/dev/tarantool/src/lib/core/coio_task.c:163:16
    #7 0x5a240fd55430 in eio_execute /home/shiny/dev/tarantool/third_party/libeio/eio.c:2015:9
    #8 0x5a240fd4f878 in etp_proc /home/shiny/dev/tarantool/third_party/libeio/etp.c:373:7
    #9 0x5a240d853156 in asan_thread_start(void*) asan_interceptors.cpp.o

SUMMARY: AddressSanitizer: 76 byte(s) leaked in 1 allocation(s).
```

Note that repro is done with #10207 applied but `coio_connect_timeout()` fixes reverted.

Introduced in `2.10.0-beta1-222-g97457a399d`.

### PR DESCRIPTION
**Pull request by nshy**
Part of #10211 

Factored out from all-in-one PR #10207

Closes #10476
Closes #10477
Closes #10479
Closes #10480
Closes #10482

### CODE REVIEW

File: src/lib/core/coio.c [hunk 1]
```diff
@@ -205,8 +210,8 @@ coio_connect_timeout(const char *host, const char *service, int host_hint,
 	diag_set(SocketError, sio_socketname(fd), "connection failed");
 	return fd;
 out:
-	if (host_hint == 0)
-		freeaddrinfo(ai);
+	if (ai_resolve != NULL)
+		freeaddrinfo(ai_resolve);
 	else
 		free(ai_local.ai_addr);
```
Comment (locker) [inline src/lib/core/coio.c:216]: Nit: Better initialize both `ai_local` and `ai_resolve` to `NULL` and free both of them here if not `NULL`.

File: changelogs/unreleased/fix-iproto-override-memleak.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+## bugfix/box
+
+- Fixed a memory leak in iproto override (introduced in 3.1).
```
Comment (locker) [inline changelogs/unreleased/fix-iproto-override-memleak.md:3]: This looks critical - the leak happens on execution of every overridden IPROTO request. Please file an issue and reference it here.

File: changelogs/unreleased/fix-replica-disconnect-memleak.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+## bugfix/replication
+
+- Fixed a memory leak on disconnection from replica.
```
Comment (locker) [inline changelogs/unreleased/fix-replica-disconnect-memleak.md:3]: I think this bug can leak quite a lot of memory if a replica connects and disconnects repeatedly due to some kind of error or poor connectivity, can't it? Please consider filing an issue for this one, as well.

File: changelogs/unreleased/fix-xlog-open-memleak.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+## bugfix/box
```
Comment (locker) [inline changelogs/unreleased/fix-xlog-open-memleak.md:1]: I'd say it's `lua`, not `box`.

File: changelogs/unreleased/fix-fkey-constraint-check-memleak.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+## bugfix/box
+
+- Fixed a memory leak on foreign key constraint check failure.
```
Comment (locker) [inline changelogs/unreleased/fix-fkey-constraint-check-memleak.md:3]: Again, this looks critical. Please file an issue.

File: src/lib/core/coio.c (hunk 1 above)
Comment (nshy) [reply src/lib/core/coio.c:216]: Done.

```diff
@ -170,7 +170,7 @@ coio_connect_timeout(const char *host, const char *service, int h>

        struct addrinfo *ai;
        struct addrinfo *ai_resolve = NULL;
-       struct addrinfo ai_local;
+       struct addrinfo ai_local = { .ai_addr = NULL };
        if (host != NULL && service != NULL && host_hint != 0) {
                if (coio_fill_addrinfo(&ai_local, host, service,
                                       host_hint) != 0)
@@ -212,7 +212,7 @@ coio_connect_timeout(const char *host, const char *service, int h>
 out:
        if (ai_resolve != NULL)
                freeaddrinfo(ai_resolve);
-       else
+       if (ai_local.ai_addr != NULL)
                free(ai_local.ai_addr);
        return fd;
```

File: changelogs/unreleased/fix-fkey-constraint-check-memleak.md (hunk 5 above)
Comment (nshy) [reply changelogs/unreleased/fix-fkey-constraint-check-memleak.md:3]: Created an issue #10476

File: changelogs/unreleased/fix-iproto-override-memleak.md (hunk 2 above)
Comment (nshy) [reply changelogs/unreleased/fix-iproto-override-memleak.md:3]: Created an issue #10477

File: changelogs/unreleased/fix-replica-disconnect-memleak.md (hunk 3 above)
Comment (nshy) [reply changelogs/unreleased/fix-replica-disconnect-memleak.md:3]: Created an issue #10480

File: changelogs/unreleased/fix-xlog-open-memleak.md (hunk 4 above)
Comment (nshy) [reply changelogs/unreleased/fix-xlog-open-memleak.md:1]: Fixed.

**Review by lenkis — APPROVED** [review]

File: changelogs/unreleased/gh-10477-fix-iproto-override-memleak.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+## bugfix/box
+
+- Fixed a memory leak in iproto override introduced in 3.1 (gh-10477).
```
Comment (locker) [inline changelogs/unreleased/gh-10477-fix-iproto-override-memleak.md:3]: - Please remove "introduced in 3.1".
- Unclear when the leak happens. Better write: "Fixed a memory leak in IPROTO on execution of the callback installed with `box.iproto.override()`".

File: changelogs/unreleased/gh-10479-fix-xlog-open-memleak.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+## bugfix/lua
+
+- Fixed a memory leak on `xlog` open failure (gh-10479).
```
Comment (locker) [inline changelogs/unreleased/gh-10479-fix-xlog-open-memleak.md:3]: Better write: "Fixed a memory leak in `xlog.pairs()` on failure to open the file".

**Review by locker — CHANGES_REQUESTED** [review]

File: changelogs/unreleased/gh-10477-fix-iproto-override-memleak.md (hunk 6 above)
Comment (nshy) [reply changelogs/unreleased/gh-10477-fix-iproto-override-memleak.md:3]: Fixed.

File: changelogs/unreleased/gh-10479-fix-xlog-open-memleak.md (hunk 7 above)
Comment (nshy) [reply changelogs/unreleased/gh-10479-fix-xlog-open-memleak.md]: Fixed.

**Review by locker — APPROVED** [review]

**Comment by locker** [comment]
Cherry-picked to 2.11 and 3.2.

**Pull request merged**